zygrader/bobs_shake.py

```diff
@@ -0,0 +1,422 @@
+        ta_netid = row[4 - old_format_shift]
+
+        uniq_item = event_type + student_name + assignment_name
+
+        # FIXME: ta_netid is a lab name if student has a comman in their name
+        # (this is somewhat fixed in data/lock.py by using csv to write data,
+        # but old data might still have commas)
+        #
+        # one possible fix (and what I'm thinking I'll do) is to modify the
+        # the lock log files outside of zygrader,
+        # just with some separate script, to keep this area clean
+
+        lock_type = row[5 - old_format_shift]
+        is_lock = lock_type == "LOCK"
+
+        return _WorkEvent(time_stamp, event_type, student_name, ta_netid,
+                          is_lock, row, uniq_item)
+
+    uniq_help_id = 0
+
+    @classmethod
+    def from_queue_data_start_and_end(cls, row):
+        try:
+            student_name = row[1]
+            ta_name = row[2]
+
+            begin_time_str = row[4]
+            begin_time = datetime.datetime.strptime(begin_time_str,
+                                                    "%m/%d/%Y %I:%M:%S %p")
+
+            duration_str = row[7]
+            if duration_str == "None":  # student helped themselves
+                return None, None
+            try:
+                minute_str, second_str = duration_str.split(":")
+                minutes = int(minute_str)
+                seconds = int(second_str)
+            except ValueError:
+                cls.queue_errors.append(row)
+                return None, None
+            duration_delta = datetime.timedelta(minutes=minutes,
+                                                seconds=seconds)
+
+            end_time = begin_time + duration_delta
+
+            uniq_item = f'HELP{cls.uniq_help_id}'
+            cls.uniq_help_id += 1
+
+            begin_event = _WorkEvent(begin_time, 'HELP', student_name, ta_name,
+                                     True, row, uniq_item)
+            end_event = _WorkEvent(end_time, 'HELP', student_name, ta_name,
+                                   False, row, uniq_item)
+
+            return begin_event, end_event
+        except Exception:
+            cls.queue_errors.append(row)
+            return None, None
+
+    def __str__(self):
+        return (f"At {self.time_stamp} {self.ta_name} "
+                f"{'began' if self.is_begin else 'finished'} "
+                f"{self.student_name}'s {self.event_type}"
+                f"({self.og_data})")
+
+
+def _deduplicate_nested_events(og_list):
+    new_list = []
+    unmatched_depth = 0
+
+    for event in og_list:
+        if unmatched_depth:
+            if event.is_begin:
+                unmatched_depth += 1
+            else:
+                unmatched_depth -= 1
+                if unmatched_depth == 0:
+                    new_list.append(event)
+        else:
+            if event.is_begin:
+                unmatched_depth += 1
+                new_list.append(event)
+            else:
+                pass
```

Comment:
Remove

Reply:
I had that to make it clear that nothing needs to happen there if the event is not begin. but a comment would probably be even better